service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/ServiceClient.java

```diff
@@ -65,11 +110,102 @@ public static ServiceClient createFromConnectionString(String connectionString,
             throw new IllegalArgumentException("ServiceClientOptions cannot be null for this constructor");
         }
 
-        // Codes_SRS_SERVICE_SDK_JAVA_SERVICECLIENT_12_002: [The constructor shall create IotHubConnectionString object using the IotHubConnectionStringBuilder]
-        IotHubConnectionString iotHubConnectionString = IotHubConnectionStringBuilder.createConnectionString(connectionString);
+        IotHubConnectionString iotHubConnectionString = IotHubConnectionStringBuilder.createIotHubConnectionString(connectionString);
 
-        // Codes_SRS_SERVICE_SDK_JAVA_SERVICECLIENT_12_003: [The constructor shall create a new instance of ServiceClient using the created IotHubConnectionString object and return with it]
-        return new ServiceClient(iotHubConnectionString, iotHubServiceClientProtocol, options);
+        TokenCredential authenticationTokenProvider = new IotHubConnectionStringCredential(connectionString);
+
+        this.hostName = iotHubConnectionString.hostName;
+        this.authenticationTokenProvider = authenticationTokenProvider;
+        this.tokenCredentialType = CbsAuthorizationType.SHARED_ACCESS_SIGNATURE;
+        this.iotHubServiceClientProtocol = iotHubServiceClientProtocol;
+        this.options = options;
+        this.amqpMessageSender =
+                new AmqpSend(
+                        hostName,
+                        authenticationTokenProvider,
+                        CbsAuthorizationType.SHARED_ACCESS_SIGNATURE,
+                        iotHubServiceClientProtocol,
+                        options.getProxyOptions(),
+                        options.getSslContext());
+    }
+
+    /**
+     * Create a {@link ServiceClient} instance with a custom {@link TokenCredential} to allow for finer grain control
+     * of authentication tokens used in the underlying connection.
+     *
+     * @param hostName The hostname of your IoT Hub instance (For instance, "your-iot-hub.azure-devices.net")
+     * @param authenticationTokenProvider The custom {@link TokenCredential} that will provide authentication tokens to
+     *                                    this library when they are needed.
+     * @param tokenCredentialType The type of authentication tokens that the provided {@link TokenCredential}
+     *                          implementation will always give.
+     * @param iotHubServiceClientProtocol The protocol to open the connection with.
+     * @return The created {@link ServiceClient} instance.
+     */
+    public ServiceClient(
+            String hostName,
+            TokenCredential authenticationTokenProvider,
+            CbsAuthorizationType tokenCredentialType,
+            IotHubServiceClientProtocol iotHubServiceClientProtocol)
```

Comment:
Did we update any of our E2E tests to use this new API?

Reply:
sample?

Reply:
I need to add some samples, but I did add a basic c2d, file upload notification receiver and feedback receiver test to the ServiceClientTests.java file